Speedup pseudoheader lookup

franz1981 opened the pull request:
Motivation:

Pseudoheader's could be cached within Hpack static table, but when they are added into Http2HeadersSink they are validated, verifying if they are pseudo-headers. Given that the AsciiStrings used within PseudoHeaderName and HpackStaticTable, it always requires to perform `AsciiString's equality.
For the same reason, if user code lookup those pseudo-headers using (as expected) PseudoHeaderName, the same equality check will always be performed.

Modification:

Speed-up pseudo-header lookup algorithm using a simple switch and ensuring that PseudoHeader's name are used to populate HpackStaticTable, to remove the need of a content equality check.

Result:

Faster pseudo-header lookup

franz1981 commented:
@normanmaurer FYI

![image](https://github.com/netty/netty/assets/13125299/d86c8e9b-05fe-4bf0-b15d-3300729eee0a)

This is just a tiny fraction of the whole cost (it's no more than 5-10% then the whole), but I'm trying hard to reduce the overhead of lookup to the bare minimum for the most used headers names.

franz1981 commented:
Currently the switch approach is about twice as fast of the existing one using the map, in pretty much any case but with normal strings, where it is just 10% faster.

The approach I have used for the pseudo headers in the static table could be reused with the other known header names (eg content-type) which are currently in HttpHeaderNames and would allow faster lookups from user code which use them, skipping a wasteful content equality and just perform a reference equality, wdyt @normanmaurer ?

This approach can guarantee a very important performance improvement which extend beyond the internals of Netty, to users (and framework built on top of it)

franz1981 commented on codec-http2/src/main/java/io/netty/handler/codec/http2/Http2Headers.java:
```diff
@@ -68,12 +68,25 @@ enum PseudoHeaderName {
+         * This has been implemented in order to allow the JIT
+         * to produce a binary-search like sequence of conditional
+         */
+        private static final int METHOD_HASH_CODE = -240705714;
```
These requires to add a version based on big Indian as well :/

Sadly ASCII hashcode uses Unsafe under the hood, which depends on endianess :(

chrisvest replied:
yeah, long outstanding bug with the endianess

franz1981 commented:
this PR push a significant change in case the Hpack static table hit, which would impact the speed of the mentioned code path at https://github.com/netty/netty/blob/ee9bb63aa5382bf9930eb0df4b690286173f126a/codec-http2/src/main/java/io/netty/handler/codec/http2/HpackDecoder.java#L559-L561 given that both 

`headers.add` and `getPseudoHeader` relies on `AsciiString.contentEquals` which would be speedup by *not* performing content equality but just performing a quick reference equality on an already present `AsciiString::hashCode`.
I've performed a similar treatment to the other known HttpHeaders.

@normanmaurer @chrisvest let me know how it looks like, if I've understood correctly what the static HPack table is used for...

normanmaurer reviewed (approved):
This looks great!

franz1981 commented on common/src/main/java/io/netty/util/AsciiString.java:
```diff
@@ -1044,7 +1044,10 @@ public boolean contentEquals(CharSequence a) {
-
+        final String asString = this.string;
+        if (asString != null) {
+            return asString.contentEquals(a);
```
This is of great help in case framework writers have a `String`-based API for the Netty http headers, allowing a bound-free check (vectorized too) to check for equality

franz1981 replied:
I have to better think about this; what if a user create a cached AsciiString value providing a non Latin String? I am opening a separate issue for this, but right now is not safe use String equality
because  https://github.com/netty/netty/blob/a5cc7a89ab411212b7fabb5a57ff765686771d59/common/src/main/java/io/netty/util/AsciiString.java#L1407 doesn't perform any validation/transformation to make the string content the same as AsciiString

franz1981 commented:
I think I could further improve if I can detect that headers/pseudo headers are "hydrated" from the static/dynamic table and save part of the validation to happen.
For the header's map insertion, it could be great to save any additional header name validation to happen for "trusted" header names (or, just obtained from hpack tables, especially static ones). Wdyt @idelpivnitskiy ?
Don't want to cause CVEs with this, eh, but I see revalidation as a waste of cycles, if that happen over and over for already validated header names.

franz1981 commented:
@chrisvest @normanmaurer I've changed my mind, and I've improved the lookup in every single case, see

before c1aca1202044a69a37adca48953b2f7b146a01b5  
```
Benchmark                                                        (hits)  (same)  Mode  Cnt    Score    Error  Units
Http2PseudoHeadersLookupBenchmark.getAsciiStringPseudoHeader       true    true  avgt   20   16.072 ±  0.080  ns/op
Http2PseudoHeadersLookupBenchmark.getAsciiStringPseudoHeader       true   false  avgt   20   43.031 ±  0.267  ns/op
Http2PseudoHeadersLookupBenchmark.getNewAsciiStringPseudoHeader    true    true  avgt   20   95.165 ± 13.537  ns/op
Http2PseudoHeadersLookupBenchmark.getNewAsciiStringPseudoHeader    true   false  avgt   20  105.938 ±  5.767  ns/op
Http2PseudoHeadersLookupBenchmark.getStringPseudoHeader            true    true  avgt   20   82.462 ±  1.555  ns/op
Http2PseudoHeadersLookupBenchmark.getStringPseudoHeader            true   false  avgt   20   82.244 ±  0.151  ns/op
```
on c1aca1202044a69a37adca48953b2f7b146a01b5
```
Benchmark                                                        (hits)  (same)  Mode  Cnt   Score    Error  Units
Http2PseudoHeadersLookupBenchmark.getAsciiStringPseudoHeader       true    true  avgt   20   7.350 ±  0.212  ns/op
Http2PseudoHeadersLookupBenchmark.getAsciiStringPseudoHeader       true   false  avgt   20  16.102 ±  0.075  ns/op
Http2PseudoHeadersLookupBenchmark.getNewAsciiStringPseudoHeader    true    true  avgt   20  38.207 ±  3.491  ns/op
Http2PseudoHeadersLookupBenchmark.getNewAsciiStringPseudoHeader    true   false  avgt   20  27.765 ± 10.314  ns/op
Http2PseudoHeadersLookupBenchmark.getStringPseudoHeader            true    true  avgt   20  23.331 ±  0.481  ns/op
Http2PseudoHeadersLookupBenchmark.getStringPseudoHeader            true   false  avgt   20  23.517 ±  0.167  ns/op
```

The use case, thanks to this PR aiming to reuse pseudoheader's `AsciiString`, which care the most, is the very first one (`getAsciiStringPseudoHeader` with `same` and `hits`), given that after the changes in this PR we reuse the same `AsciiString` of the pseudo header name's enum, but we're mildly interested in the rest.
The `hits = false` case is not interesting (unelss for CVEs), because HTTP 2 validators prevent using others pseudoheader(s).

The `String`-based uses cases are interesting in case users uses APIs to query pseudoheader(s) which make uses of `String` literals, conversely they're gonna pay the cost of an intrinsified `String::hashCode` which should be still decent.

I'm now adding a further commit to fix the java 1.6 compatibility issue at
```
Error:  Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.8.0:compile (default-compile) on project netty-codec-http2: Compilation failure
Error:  /home/runner/work/netty/netty/codec-http2/src/main/java/io/netty/handler/codec/http2/Http2Headers.java:[269,19] error: strings in switch are not supported in -source 1.6
Error:    (use -source 7 or higher to enable strings in switch)
Error:  -> [Help 1]
Error:  
Error:  To see the full stack trace of the errors, re-run Maven with the -e switch.
Error:  Re-run Maven using the -X switch to enable full debug logging.
Error:  
Error:  For more information about the errors and possible solutions, please read the following articles:
Error:  [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoFailureException
Error:  
Error:  After correcting the problems, you can resume the build with the command
Error:    mvn <args> -rf :netty-codec-http2
Error: Process completed with exit code 1.
```
Which is easy to be solved (with some ugly code, as usual).

normanmaurer commented on codec-http2/src/main/java/io/netty/handler/codec/http2/Http2Headers.java:
```diff
@@ -113,7 +120,197 @@ public static boolean isPseudoHeader(CharSequence header) {
+        private static final int STATUS_STRING_HASHCODE = -956875604;
+        private static final int PATH_STRING_HASHCODE = 56997727;
+        private static final int AUTHORITY_STRING_HASHCODE = -1332238263;
+        private static final int PROTOCOL_STRING_HASHCODE = 1552659666;
```
Why not get these values in a static block instead of hard coding these ?

franz1981 replied:
Javac need to see then as constant, in order to use them in the switch cases :/ 
But String::hashCode is a fixed contractor, hence shouldn't be a big deal

franz1981 replied:
But is ugly, nonetheless:"(

franz1981 replied:
what's your final thought on this bud?

normanmaurer replied:
ok I think its fine... maybe add some more comments ?

normanmaurer commented:
@idelpivnitskiy @chrisvest PTAL

franz1981 commented on codec-http2/src/main/java/io/netty/handler/codec/http2/Http2Headers.java:
```diff
@@ -113,7 +120,197 @@ public static boolean isPseudoHeader(CharSequence header) {
+        static {
+            // do the same but throw a runtime exception here
+            // this is to ensure that the hashcode of the strings are the same as the hashcode of the enum
+            if (METHOD.value().toString().hashCode() != METHOD_STRING_HASHCODE) {
```
It's a good idea, just wanna make sure that will be noisy at least for our test suite, where -ea should be enabled (it is?)

normanmaurer replied:
yep

normanmaurer replied:
maybe what we could do is expose these as package private and add something to the tests so we would have a failing unit test ?

normanmaurer commented on microbench/src/main/java/io/netty/handler/codec/http2/Http2PseudoHeadersLookupBenchmark.java:
```diff
@@ -0,0 +1,114 @@
+ */
+
+/*
+ * Copyright 2015 Twitter, Inc.
```
why this header ?

franz1981 replied:
I belvive Is my GitHub copilot!!! Ahahahah just I don't understand why twitter, as well...

normanmaurer commented on codec-http2/src/test/java/io/netty/handler/codec/http2/Http2HeadersTest.java:
```diff
@@ -0,0 +1,83 @@
+        assertEquals(new String(STATUS.toString().toCharArray()), STATUS_STRING_HASHCODE);
+        assertEquals(new String(PATH.toString().toCharArray()), PATH_STRING_HASHCODE);
+        assertEquals(new String(AUTHORITY.toString().toCharArray()), AUTHORITY_STRING_HASHCODE);
+        assertEquals(new String(PROTOCOL.toString().toCharArray()), PROTOCOL_STRING_HASHCODE);
```
@franz1981 I think you missed to call `hashCode()` on these ?

franz1981 replied:
Oh gosh, you are right :/

normanmaurer commented:
@idelpivnitskiy @chrisvest PTAL

normanmaurer commented:
@bryce-anderson PTAL

bryce-anderson commented on codec-http2/src/main/java/io/netty/handler/codec/http2/Http2Headers.java:
```diff
@@ -113,7 +120,194 @@ public static boolean isPseudoHeader(CharSequence header) {
+                        return PATH;
+                    }
+                    // :path
+                    return getPathIfEqualsNoPrefix(array, offset);
```
How often will these specialized paths be used? In my minds eye we can get the majority of the benefit from a `.equals(..)` and maybe some improvements to `AsciiString.equals(..)`.
We could probably speed up `AsciiString.equals(..)` a little bit. We could start off with a reference equality check. Second, the terminal `&&` sequence could skip the hash code checks if the hash code hasn't been computed: it's intended to be a short circuit to avoid iteration but if the hash code needs to be computed anyway it's no benefit and actually a big detriment since to compute the hash code we have to traverse the entire array. The `AsciiString.hashCode(..)` computation looks like it was a dominator in your flame graph.

normanmaurer replied:
I like the idea of @bryce-anderson

franz1981 replied:
> How often will these specialized paths be used?

If you mean the content equality ones, in 2 cases:
1. malformed/fake pseudo-headers eg `:patc` this will go through `getPathIfEqualsNoPrefix` and similar
2. non-constant known pseudo-header(s) eg new AsciiString(":path"), will still hit `getPathIfEqualsNoPrefix` and similar

Both seems unlikely path, or at least, with this PR the last one shouldn't be "hot" anymore.
I don't remember by heart the Http 2 spec but AFAIK they should be populated by using the static table only, surely @normanmaurer remember...

> We could probably speed up AsciiString.equals(..) a little bit.

I will look at it as well, based on these insights, thanks!
Related what I could do on the slow-paths is to simplify it using `AsciiString`'s equals calls still retaining the ref equality fast paths out of such, in case of length-collisions. If hash code would happen...it is what it is, see below...

> if the hash code needs to be computed anyway it's no benefit and actually a big detriment since to compute the hash code we have to traverse the entire array.

Right, I'm aware that the existing code using `getPseudoheader` will either do it because of:
- Netty internal header name validation on HPack decoding (which happen after already inserting it on http headers)
- Netty internal header name validation on Http Header insertion (both user code and on HPack decoding see https://github.com/netty/netty/blob/ee9bb63aa5382bf9930eb0df4b690286173f126a/codec-http2/src/main/java/io/netty/handler/codec/http2/HpackDecoder.java#L559)

In both cases the `AsciiString's` hash code won't be "thrown away", so maybe I could just ignore that it happens, and re-set my expectations for the slow-paths performance (which I shouldn't care that much, effectively, unless I'm very wrong related the HTTP spec).

bryce-anderson commented on codec-http2/src/main/java/io/netty/handler/codec/http2/Http2Headers.java:
```diff
@@ -113,7 +120,194 @@ public static boolean isPseudoHeader(CharSequence header) {
+            // String::hashCode is intrinsified
+            // the cases uses String's literal which are the same used in the enums
+            // String::hashCode is a contract which is not going to change
+            switch (header.hashCode()) {
```
How much benefit is this path over the `getPseudoHeaderName(header)` call above? Having symmetry with the `AsciiString` version above would make this easier to follow and cut out a lot of extract code.

franz1981 replied:
I can re-run the experiment but i can tell you already which benefits it have:
1. given that are separate methods (without reusing the existing code for AsciiString) there's no bimorphism involved around `CharSequence`, which usually mess-up with GraalVM native image, for example
2. the string's hash-code have no collisions among each others, while using the length of string(s), yep

In term of performance, I didn't designed the benchmark to pollute the type profiling of the these methods, but it is easy to do, so I can prove what happen with OpenJDK C2 as well, if you think is relevant.

The `getMethodSchemeStatusIfEqualsNoPrefix` is what happen in  case of worst case matches, which is not great, if compared to what we can have now with `String`s and distinct hashcodes.

bryce-anderson replied:
Sorry, I wasn't clear. My suggestion was only regarding using the hashcode instead of switching on length. The benchmark of a String without a populated hashcode wasn't tested, although that starts to come back to the point you made earlier that we're going to be computing a hashcode anyway. That kind of suggests that the benchmarking should be done at the level of adding a header to the header map: this change might only matter in that context.
I'm just coming from a position that we have a lot of code here that has the same end goal but takes different approaches which is odd to me. The switch on length seems "good enough" and should work for all cases.

bryce-anderson reviewed (commented):
It looks correct to me.
I like ditching the hash table lookup but the hash-code pre-computations feel a bit excessive and now have an asymmetry between `String` and `AsciiString` which hurts readability imo, but that's obviously subjective.

franz1981 commented:
These are the 3 benchmark results (from https://github.com/netty/netty/pull/13747#issuecomment-1879736797):
before c1aca1202044a69a37adca48953b2f7b146a01b5  
```
Benchmark                                                        (hits)  (same)  Mode  Cnt    Score    Error  Units
Http2PseudoHeadersLookupBenchmark.getAsciiStringPseudoHeader       true    true  avgt   20   16.072 ±  0.080  ns/op
Http2PseudoHeadersLookupBenchmark.getAsciiStringPseudoHeader       true   false  avgt   20   43.031 ±  0.267  ns/op
Http2PseudoHeadersLookupBenchmark.getNewAsciiStringPseudoHeader    true    true  avgt   20   95.165 ± 13.537  ns/op
Http2PseudoHeadersLookupBenchmark.getNewAsciiStringPseudoHeader    true   false  avgt   20  105.938 ±  5.767  ns/op
Http2PseudoHeadersLookupBenchmark.getStringPseudoHeader            true    true  avgt   20   82.462 ±  1.555  ns/op
Http2PseudoHeadersLookupBenchmark.getStringPseudoHeader            true   false  avgt   20   82.244 ±  0.151  ns/op
```
on c1aca1202044a69a37adca48953b2f7b146a01b5
```
Benchmark                                                        (hits)  (same)  Mode  Cnt   Score    Error  Units
Http2PseudoHeadersLookupBenchmark.getAsciiStringPseudoHeader       true    true  avgt   20   7.350 ±  0.212  ns/op
Http2PseudoHeadersLookupBenchmark.getAsciiStringPseudoHeader       true   false  avgt   20  16.102 ±  0.075  ns/op
Http2PseudoHeadersLookupBenchmark.getNewAsciiStringPseudoHeader    true    true  avgt   20  38.207 ±  3.491  ns/op
Http2PseudoHeadersLookupBenchmark.getNewAsciiStringPseudoHeader    true   false  avgt   20  27.765 ± 10.314  ns/op
Http2PseudoHeadersLookupBenchmark.getStringPseudoHeader            true    true  avgt   20  23.331 ±  0.481  ns/op
Http2PseudoHeadersLookupBenchmark.getStringPseudoHeader            true   false  avgt   20  23.517 ±  0.167  ns/op
```
and the last one is, by using your suggestions (if I got them right) (at f0da758ce15edc81405707f4b8ac56a01dcd5855)
```
Benchmark                                                        (hits)  (same)  Mode  Cnt   Score   Error  Units
Http2PseudoHeadersLookupBenchmark.getAsciiStringPseudoHeader       true    true  avgt   20   8.242 ± 0.466  ns/op
Http2PseudoHeadersLookupBenchmark.getAsciiStringPseudoHeader       true   false  avgt   20  26.543 ± 0.056  ns/op
Http2PseudoHeadersLookupBenchmark.getNewAsciiStringPseudoHeader    true    true  avgt   20  56.179 ± 5.559  ns/op
Http2PseudoHeadersLookupBenchmark.getNewAsciiStringPseudoHeader    true   false  avgt   20  55.942 ± 0.624  ns/op
Http2PseudoHeadersLookupBenchmark.getStringPseudoHeader            true    true  avgt   20  20.651 ± 0.114  ns/op
Http2PseudoHeadersLookupBenchmark.getStringPseudoHeader            true   false  avgt   20  20.424 ± 0.066  ns/op
```

the `String`s suggestion seems to work very good (even better, actually, because we avoid the hash), but:
1. `getNewAsciiStringPseudoHeader` which force creating a new `AsciiString`  (which can contains the same byte array of a pseudo-header, or a copy of it) to make it compute hash code, forcibly, is receiving a performance penaltly, as expected, although not as bad as it was in `4.1` now.
2. `getAsciiStringPseudoHeader` which is using a cached `AsciiString` (ie which hashcode is already computed), and is a pseudo-header despite not being the same (ref equality) of any existing, it has received a perf penality, meaning that saving the char-by-char equality (eg https://github.com/netty/netty/blob/88ac731d6195c40af1903d9c777d1ce9882b8955/codec-http2/src/main/java/io/netty/handler/codec/http2/Http2Headers.java#L179-L196) in favour of using `AsciiString::equals` didn't work as good, could be either due to array bound checking, increased number of comparisons or because we don't split the equality checks based the uncommon letters first.


Both cases looks to me like "uncommon" expected slow paths, especially after this PR, but none prevent users to use different cached `AsciiString`s without using the existing ones, unless (@normanmaurer please confirm/reject)
HPack could skip using the static table for pseudo-header(s), meaning that we risk to fall into both above cases.
If that's the case, I would reconsider improving things, by using the char-by-char strategy.

bryce-anderson commented:
f0da758 is really close to what I was suggesting but may still suffer from the unnecessary computation of hash-codes which is really an artifact of `AsciiString.equals(..)`. Also, unnecessary may not be all that unnecessary since as you pointed out this is part of a larger operation of adding the header to a hashmap and therefore we'll compute the hash regardless for the hot path.

This patch to f0da758 is what I mean:
```
diff --git a/codec-http2/src/main/java/io/netty/handler/codec/http2/Http2Headers.java b/codec-http2/src/main/java/io/netty/handler/codec/http2/Http2Headers.java
index a2b214f952..c554ddd1be 100644
--- a/codec-http2/src/main/java/io/netty/handler/codec/http2/Http2Headers.java
+++ b/codec-http2/src/main/java/io/netty/handler/codec/http2/Http2Headers.java
@@ -134,15 +134,6 @@ public interface Http2Headers extends Headers<CharSequence, CharSequence, Http2H
                     return ":path".contentEquals(header)? PATH : null;
                 case 7:
                     // :method, :scheme, :status
-                    if (":method" == header) {
-                        return METHOD;
-                    }
-                    if (":scheme" == header) {
-                        return SCHEME;
-                    }
-                    if (":status" == header) {
-                        return STATUS;
-                    }
                     if (":method".contentEquals(header)) {
                         return METHOD;
                     }
@@ -177,16 +168,6 @@ public interface Http2Headers extends Headers<CharSequence, CharSequence, Http2H
                     // :path
                     return PATH.value().equals(header) ? PATH : null;
                 case 7:
-                    if (header == METHOD.value()) {
-                        return METHOD;
-                    }
-                    if (header == SCHEME.value()) {
-                        return SCHEME;
-                    }
-                    if (header == STATUS.value()) {
-                        return STATUS;
-                    }
-                    // :method, :scheme, :status
                     if (METHOD.value().equals(header)) {
                         return METHOD;
                     }
diff --git a/common/src/main/java/io/netty/util/AsciiString.java b/common/src/main/java/io/netty/util/AsciiString.java
index da06a1dc49..997e8e0c4a 100644
--- a/common/src/main/java/io/netty/util/AsciiString.java
+++ b/common/src/main/java/io/netty/util/AsciiString.java
@@ -1136,16 +1136,16 @@ public final class AsciiString implements CharSequence, Comparable<CharSequence>
 
     @Override
     public boolean equals(Object obj) {
-        if (obj == null || obj.getClass() != AsciiString.class) {
-            return false;
-        }
         if (this == obj) {
             return true;
         }
+        if (obj == null || obj.getClass() != AsciiString.class) {
+            return false;
+        }
 
         AsciiString other = (AsciiString) obj;
         return length() == other.length() &&
-               hashCode() == other.hashCode() &&
+                (hash == 0 || other.hash == 0 || hash == other.hash) && // don't trigger hash code computation: the follow up equality check is still necessary either way so checking hash is only a shortcut for the non-equal case.
                PlatformDependent.equals(array(), arrayOffset(), other.array(), other.arrayOffset(), length());
     }
 
```
I dropped the reference equality checks since they should happen anyway if we get lucky and the `.equals(..)` call gets inlined but it's hard to say without running it.

franz1981 commented:
@bryce-anderson 

re the patch, it's not good to avoid the 3 ref equalilty fast-path eg
1. `getPseudoHeader(":status")` is called
2. `":method".equals(":status")` would perform the ref equality check AND fail at the second iteration char (`m` != `s`); the equality could be performed by using a vectorized intrinsics, but I think, due to the string length to NOT benefit of it, and just use byte per byte comparisons (with bound checks) - i will take a look anyway
3. `":scheme".equals(":status")` will do the same, but failing at the third char (meaning it has to perform, again, 3 chars comparison AND the ref-equality check)
4. finally `":status".equals(":status")` win the ref-equality check and complete

In total, forgetting any optimization, we have to perform:
- 3 ref checks, where 1 has succeeded
- 2 + 3 chars comparisons and potential equality calls which could NOT be inlined for different reasons (eg "already compiled into a big method" or just an unfortunate inlining depth)

The optimized version, previously done, instead, just perform:
- 3 ref checks, where 1 has succeeded

I'm running a benchmark run with your suggestion just on `AsciiString`, anyway, although the microbenchmark is the best case scenario.

And it doesn't re

franz1981 commented:
Applying your full patch I got

```
Benchmark                                                        (hits)  (same)  Mode  Cnt   Score   Error  Units
Http2PseudoHeadersLookupBenchmark.getAsciiStringPseudoHeader       true    true  avgt   20  18.340 ± 0.160  ns/op
Http2PseudoHeadersLookupBenchmark.getAsciiStringPseudoHeader       true   false  avgt   20  30.618 ± 0.383  ns/op
Http2PseudoHeadersLookupBenchmark.getNewAsciiStringPseudoHeader    true    true  avgt   20  17.721 ± 0.310  ns/op
Http2PseudoHeadersLookupBenchmark.getNewAsciiStringPseudoHeader    true   false  avgt   20  27.735 ± 0.382  ns/op
Http2PseudoHeadersLookupBenchmark.getStringPseudoHeader            true    true  avgt   20  20.201 ± 0.958  ns/op
Http2PseudoHeadersLookupBenchmark.getStringPseudoHeader            true   false  avgt   20  21.198 ± 0.536  ns/op
```
which show a regression in `getAsciiStringPseudoHeader` which I would like to save, so I'm creating a new branch picking the parts which i think are acceptable as a performance degradation, and will report it here.

bryce-anderson commented:
@franz1981 lots of interesting data. I'll admit I'm biased toward readability when we're talking about differences of about 10 ns but thats obviously a subjective opinion. I'll leave it to you, @normanmaurer, and anybody else to make a final decision. All options presented so far look _correct_ to me which is the most important thing.

Regardless of what ya'll decided to do, this is a very detailed performance analysis which is pretty cool.

franz1981 commented:
Thanks @bryce-anderson for completeness I can tell that these values at my previous comment

```
Http2PseudoHeadersLookupBenchmark.getNewAsciiStringPseudoHeader    true    true  avgt   20  17.721 ± 0.310  ns/op
Http2PseudoHeadersLookupBenchmark.getNewAsciiStringPseudoHeader    true   false  avgt   20  27.735 ± 0.382  ns/op
```
related your full patch doesn't seem correct and despite i have a very stable box (isolcpu/numactl isolated/turbo boost off) has produced just once that values, which is not fun...

Anyway, exactly like you, IDK what could be done best although I could send a final commit which try to get the best of both worlds

normanmaurer commented:
@franz1981 yeah lets do a final commit with best of both worlds please

franz1981 commented:
In the next days @normanmaurer I promise we can close this :P 
But happy about the interaction with @bryce-anderson, thanks again 🙏

normanmaurer commented:
@franz1981 thanks... just ping me when ready

franz1981 commented:
Ready to go @normanmaurer and @bryce-anderson 

some numbers

I've modified the benchmark to include a case which have the insertion in the headers map as well, which should perform a validation over the headers etc etc to emulate what HPack does, but specifically for pseudoheaders.

I've removed the benchmark cases which we allocating strings/asciistring just for the sake of forcing hash code computation: this is the slowest path we try hard to avoid, with this PR, so I see no points into checking it.


The baseline on 4.1 looks like:
```
Benchmark                                                              (same)  Mode  Cnt    Score    Error  Units
Http2PseudoHeadersLookupBenchmark.addAsciiStringInHttp2Headers           true  avgt   20  144.130 ± 12.018  ns/op
Http2PseudoHeadersLookupBenchmark.addAsciiStringInHttp2Headers          false  avgt   20  191.564 ±  8.842  ns/op
Http2PseudoHeadersLookupBenchmark.addStringInHttp2Headers                true  avgt   20  352.912 ± 12.620  ns/op
Http2PseudoHeadersLookupBenchmark.addStringInHttp2Headers               false  avgt   20  337.322 ± 15.688  ns/op
Http2PseudoHeadersLookupBenchmark.getAsciiStringPseudoHeader             true  avgt   20   19.153 ±  0.227  ns/op
Http2PseudoHeadersLookupBenchmark.getAsciiStringPseudoHeader            false  avgt   20   42.377 ±  0.120  ns/op
Http2PseudoHeadersLookupBenchmark.getStringPseudoHeader                  true  avgt   20   82.420 ±  0.087  ns/op
Http2PseudoHeadersLookupBenchmark.getStringPseudoHeader                 false  avgt   20   82.356 ±  0.040  ns/op
```

While 7cc21953e455c1910e8dea086de174d932ae3ce7 :
```
Benchmark                                                              (same)  Mode  Cnt    Score    Error  Units
Http2PseudoHeadersLookupBenchmark.addAsciiStringInHttp2Headers           true  avgt   20  128.374 ± 11.856  ns/op
Http2PseudoHeadersLookupBenchmark.addAsciiStringInHttp2Headers          false  avgt   20  186.544 ± 11.011  ns/op
Http2PseudoHeadersLookupBenchmark.addStringInHttp2Headers                true  avgt   20  192.007 ±  8.835  ns/op
Http2PseudoHeadersLookupBenchmark.addStringInHttp2Headers               false  avgt   20  224.983 ± 11.036  ns/op
Http2PseudoHeadersLookupBenchmark.getAsciiStringPseudoHeader             true  avgt   20    8.630 ±  0.054  ns/op
Http2PseudoHeadersLookupBenchmark.getAsciiStringPseudoHeader            false  avgt   20   26.566 ±  0.016  ns/op
Http2PseudoHeadersLookupBenchmark.getStringPseudoHeader                  true  avgt   20   10.036 ±  0.098  ns/op
Http2PseudoHeadersLookupBenchmark.getStringPseudoHeader                 false  avgt   20   20.590 ±  0.281  ns/op
```
It's clear that "in the global picture" of adding to the map (which include removing as well, so it's not completely fair), the improvement still exists but is way more interesting for `String`s vs `AsciiString`s, despite the former get an hit while added to the header's map (which is not completely fair, given that HPack decoders will never do that!), because their ascii hash code cannot be cached.

The most interesting perrformance increase IMO, is that we now always hit the fastest path in the real world ie HPack decoded `AsciiString`s in the static table are the same as the pseudo-headers constants AND users could use them too, to perform the fastest lookup/insertion into headers map.

the code complaxity has been further reduced thanks to @bryce-anderson suggestions, although not the possible peak performence, but as said, in the global picture involving maps operations, is just matter of a bunch of nanoseconds.

normanmaurer commented on codec-http2/src/main/java/io/netty/handler/codec/http2/Http2Headers.java:
```diff
@@ -113,7 +120,89 @@ public static boolean isPseudoHeader(CharSequence header) {
+        }
+
+        private static PseudoHeaderName getPseudoHeaderName(CharSequence header) {
+            if (header.length() > 0 && header.charAt(0) == PSEUDO_HEADER_PREFIX) {
+                switch (header.length()) {
```
```suggestion
            int length = header.length();
            if (length > 0 && header.charAt(0) == PSEUDO_HEADER_PREFIX) {
                switch (length) {
```

normanmaurer commented on codec-http2/src/test/java/io/netty/handler/codec/http2/Http2HeadersTest.java:
```diff
@@ -0,0 +1,61 @@
+/*
+ * Copyright 2023 The Netty Project
```
```suggestion
 * Copyright 2024 The Netty Project
```

normanmaurer commented on microbench/src/main/java/io/netty/handler/codec/http2/Http2PseudoHeadersLookupBenchmark.java:
```diff
@@ -0,0 +1,121 @@
+/*
+ * Copyright 2023 The Netty Project
```
```suggestion
 * Copyright 2024 The Netty Project
```

normanmaurer reviewed (changes requested):
Almost there... I really like how this turned out! Great work @franz1981 and @bryce-anderson

franz1981 commented:
done @normanmaurer ;)

normanmaurer commented:
@franz1981 thanks a lot for all the work on this!
